hikari/impl/interaction_server.py

```diff
@@ -0,0 +1,558 @@
+
+    Parameters
+    ----------
+    entity_factory : hikari.api.entity_factory.EntityFactory
+        The entity factory instance this server should use.
+    event_factory : hikari.api.event_factory.EventFactory
+        The event factory instance this server should use.
+
+    Other Parameters
+    ----------------
+    dumps : aiohttp.typedefs.JSONEncoder
+        The JSON encoder this server should use. Defaults to `json.dumps`.
+    loads : aiohttp.typedefs.JSONDecoder
+        The JSON decoder this server should use. Defaults to `json.loads`.
+    public_key : builtins.bytes
+        The public key this server should use for verifying request payloads from
+        Discord. If left as `builtins.None` then the client will try to work this
+        out using `rest_client`.
+    rest_client : hikari.api.rest.RESTClient
+        The client this should use for making REST requests.
+    """
+
+    __slots__: typing.Sequence[str] = (
+        "_application_fetch_lock",
+        "_dumps",
+        "_entity_factory",
+        "_event_factory",
+        "_future",
+        "_hashes",
+        "_listeners",
+        "_loads",
+        "_rest_client",
+        "_runner",
+        "_server",
+        "_verify",
+    )
+
+    def __init__(
+        self,
+        *,
+        dumps: aiohttp.typedefs.JSONEncoder = data_binding.dump_json,
+        entity_factory: entity_factory_.EntityFactory,
+        event_factory: event_factory_.EventFactory,
+        loads: aiohttp.typedefs.JSONDecoder = data_binding.load_json,
+        rest_client: rest_client_.RESTClient,
+        public_key: typing.Optional[bytes] = None,
+    ) -> None:
+        self._application_fetch_lock = asyncio.Lock()
+        self._dumps = dumps
+        self._entity_factory = entity_factory
+        self._event_factory = event_factory
+        self._future: asyncio.Future[None] = asyncio.Future()
+        self._listeners: ListenerDictT[bases.PartialInteraction, special_endpoints.InteractionResponseBuilder] = {}
+        self._loads = loads
+        self._rest_client = rest_client
+        self._runner: typing.Optional[aiohttp.web_runner.AppRunner] = None
+        self._server = aiohttp.web.Application()
+        self._server.add_routes([aiohttp.web.post("/", self.aiohttp_hook)])
+        self._verify = ed25519.build_ed25519_verifier(public_key) if public_key is not None else None
+
+    @property
+    def is_alive(self) -> bool:
+        return self._runner is not None
+
+    async def _fetch_public_key(self) -> ed25519.VerifierT:
+        application: typing.Union[applications.Application, applications.AuthorizationApplication]
+        async with self._application_fetch_lock:
+            if self._verify:
+                return self._verify
+
+            try:
+                application = (await self._rest_client.fetch_authorization()).application
+
+            except errors.UnauthorizedError:
+                application = await self._rest_client.fetch_application()
```

Comment:
One only works with bearer authorization, one only works with bot authorization; there's zero overlap between them

Reply:
Yep, that is what I was missing :)

Reply:
Also, something that came to mind. Wouldn't it be better to check what token we were provided and to the action based on that?

Reply:
> Because, at the moment, we might be raising UnauthorizedErrors every time we need the application without any need, which might lead to Discord bans.

It makes two requests max for the whole runtime of a program, that's not gonna be an issue.

> Also, something that came to mind. Wouldn't it be better to check what token we were provided and to the action based on that? 

We just don't have access to that information here.

Reply:
Unresolving this as it can be changed once #627 is merged